ngraph/core/src/op/matmul.cpp

```diff
@@ -113,68 +53,152 @@ shared_ptr<Node> op::MatMul::clone_with_new_inputs(const OutputVector& new_args)
         }
 
-        if (arg0_rank == 1 && arg1_rank == 1)
+        // 2. One-dimensional tensors unsqueezing is applied to each input independently.
+        if (arg0_rank == 1)
         {
-            NGRAPH_CHECK(arg0_shape_update == arg1_shape_update, "Incompatible arg shapes");
-            output_shape = Shape{};
+            // If the first input is 1D tensor, it is unsqueezed to 2D tensor (row vector)
+            // by adding axes with size 1 at ROW_INDEX_DIM, to the left of the shape.
+            // For example {S} will be reshaped to {1, S}.
+            arg0_shape_tmp.insert(arg0_shape_tmp.begin(), 1);
+            arg0_rank = arg0_shape_tmp.size();
         }
-        else if (arg0_rank == 1)
+        if (arg1_rank == 1)
         {
-            // i.e., arg0 shape {3}, arg1 shape{2, 3, 2}, output shape {2, 2}
-            NGRAPH_CHECK(arg0_shape_update[0] == arg1_shape_update[arg1_rank - 2],
-                         "Incompatible arg shapes");
-            arg1_shape_update.erase(arg1_shape_update.begin() + arg1_rank - 2);
-            output_shape = arg1_shape_update;
+            // If the second input is 1D tensor, it is unsqueezed to 2D tensor (column vector)
+            // by adding axes with size 1 at COL_INDEX_DIM, to the right of the shape.
+            // For example {S} will be reshaped to {S, 1}.
+            arg1_shape_tmp.insert(arg1_shape_tmp.end(), 1);
+            arg1_rank = arg1_shape_tmp.size();
         }
-        else if (arg1_rank == 1)
+
+        // Check matrices dimensions compatibility,
+        // COL_INDEX_DIM of the first matrix has to match ROW_INDEX_DIM of the second matrix.
+        auto merged_dimension = Dimension::dynamic();
+        auto arg0_col_dim = arg0_shape_tmp[arg0_rank - 1];
+        auto arg1_row_dim = arg1_shape_tmp[arg1_rank - 2];
+        NGRAPH_CHECK(Dimension::merge(merged_dimension, arg0_col_dim, arg1_row_dim) ||
+                         arg0_col_dim.is_dynamic() || arg1_row_dim.is_dynamic(),
+                     "Incompatible MatMul matrix dimension. ",
+                     "First input dimension=",
+                     arg0_col_dim,
+                     " at COL_INDEX_DIM=",
+                     (arg0_rank - 1),
+                     " doesn't match the second input dimension=",
+                     arg1_row_dim,
+                     " at ROW_INDEX_DIM=",
+                     (arg1_rank - 2));
+
+        // 3. If ranks of input arguments are different after steps 1 and 2,
+        // the smaller tensor is unsqueezed from the left side of the shape
+        // by necessary number of axes to make both shapes of the same rank.
+        std::vector<Dimension> small_batch_matrix =
+            arg0_rank > arg1_rank ? arg1_shape_tmp : arg0_shape_tmp;
+        std::vector<Dimension> big_batch_matrix =
+            arg0_rank > arg1_rank ? arg0_shape_tmp : arg1_shape_tmp;
+        std::vector<Dimension> output_shape(big_batch_matrix);
```

Comment:
Please move this block under the next condition and modify it because these manipulations are required only when arg0_rank != arg1_rank. Introducing new small/big batch matrix variables look to be a waste of time. Insert 1's directly to one of arg0_shape_tmp or arg1_shape_tmp depending on the ranks comparison. Define output_shape after the condition when both vectors are of the same size.

Reply:
Done, I think this part looks much better now. Thanks.